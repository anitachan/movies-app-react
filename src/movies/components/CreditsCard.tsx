import { Box, Button, Typography } from '@mui/material';
import noUser from '../../assets/no-user.png';
import { ActorCast } from '../models/actorCast.model';
import { useNavigate } from 'react-router-dom';

interface Props {
	credit: ActorCast;
}

export const CreditsCard = ({ credit }: Props) => {
	const url = process.env.REACT_APP_IMAGES_URL;
	const posterSrc = credit.poster_path ? `${url}${credit.poster_path}` : noUser;
	const navigate = useNavigate();

	const handleClick = () => {
		navigate(`/details/${credit.id!}`);
	};

	return (
		<Button onClick={handleClick}>
			<Box sx={{ width: 150, textAlign: 'center' }}>
				<Box
					component="img"
					src={posterSrc}
					alt={credit.title}
					loading="lazy"
					onError={(e) => {
						e.currentTarget.onerror = null;
						e.currentTarget.src = noUser;
					}}
					sx={{
						width: '100%',
						height: 225, // ratio 2:3
						objectFit: 'cover',
						borderRadius: 1,
						display: 'block',
					}}
				/>
				<Typography variant="subtitle2" noWrap>
					{credit.title}
				</Typography>
				<Typography variant="caption" color="text.secondary" noWrap>
					{credit.release_date ? new Date(credit.release_date).getFullYear() : 'â€”'} as {credit.character}
				</Typography>
			</Box>
		</Button>
	);
};
