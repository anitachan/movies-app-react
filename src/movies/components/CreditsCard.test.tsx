import { fireEvent, render, screen } from '@testing-library/react';
import { CreditsCard } from './CreditsCard';
import { fixtureOf } from 'ts-mock-autofixture-kit';
import { ActorCast } from '../models/actorCast.model';
import noUser from '../../assets/no-user.png';

const mockUseNavigate = jest.fn();
jest.mock(
  'react-router-dom',
  () => ({
    useNavigate: () => mockUseNavigate,
  }),
  { virtual: true },
);

describe('CreditsCard', () => {
  const mockCredit = fixtureOf<ActorCast>({ seed: 42 }).create();

  test('CreditsCard matches snapshot', () => {
    const { container } = render(<CreditsCard credit={mockCredit} />);

    expect(container).toMatchSnapshot();
  });

  test('should redirect when see more is clicked', () => {
    render(<CreditsCard credit={mockCredit} />);
    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(mockUseNavigate).toHaveBeenCalledWith(`/details/${mockCredit.id!}`);
  });

  test('should set noUser when poster is empty', () => {
    const mockCreditPoster = fixtureOf<ActorCast>()
      .with('release_date', () => new Date().toISOString())
      .with('poster_path', () => null)
      .with('title', () => 'No Poster Movie')
      .create();

    render(<CreditsCard credit={mockCreditPoster} />);
    const image = screen.getByAltText('No Poster Movie') as HTMLImageElement;

    expect(image).toBeInTheDocument();
    expect(image.src).toContain(noUser);
  });

  test('should set an empty string when release date is empty', () => {
    const mockCreditReleaseDate = fixtureOf<ActorCast>()
      .with('release_date', () => '')
      .create();
    render(<CreditsCard credit={mockCreditReleaseDate} />);

    const elem = screen.getByTestId('cast-credit-info');
    expect(elem).toHaveTextContent(`â€” as ${mockCreditReleaseDate.character}`);
  });

  test('should set no user when image has an error ', () => {
    render(<CreditsCard credit={mockCredit} />);
    const img = screen.getByAltText(mockCredit.title) as HTMLImageElement;

    expect(img.src).toContain(mockCredit.poster_path);

    fireEvent.error(img);
    expect(img.src).toContain('no-user.png');
  });
});
