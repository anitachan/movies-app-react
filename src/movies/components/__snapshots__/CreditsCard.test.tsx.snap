// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditsCard CreditsCard matches snapshot 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1uent87-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <div
      class="MuiBox-root css-1ug4zxe"
    >
      <img
        alt="3CpplugYH0Te"
        class="MuiBox-root css-17yky46"
        loading="lazy"
        src="https://image.tmdb.org/t/p/originalaGTqxm8FGpgY"
      />
      <h6
        class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-noWrap css-13k9d3g-MuiTypography-root"
      >
        3CpplugYH0Te
      </h6>
      <span
        class="MuiTypography-root MuiTypography-caption MuiTypography-noWrap css-fgvunj-MuiTypography-root"
        data-testid="cast-credit-info"
      >
        2025
         as
         
        i1CDBoab94WJ
      </span>
    </div>
  </button>
</div>
`;
